### PR TITLE
Feature/expiring tags

### PR DESCRIPTION
**Pull request by chrisrecalis**

### CODE REVIEW

File: core/src/main/kotlin/io/rover/campaigns/core/events/Interfaces.kt [hunk 1]
```diff
@@ -63,7 +63,11 @@ interface EventQueueServiceInterface {
      *
      * Asynchronous, will immediately return.
      */
-    fun trackScreenViewed(screenName: String, contentID: String? = null, contentName: String? = null)
+    fun trackScreenViewed(
+        screenName: String,
+        contentID: String? = null,
+        contentName: String? = null
+    )
```
Comment (chrisrecalis) [inline core/src/main/kotlin/io/rover/campaigns/core/events/Interfaces.kt:70]: IDE formatting did this. can revert if need be

File: core/src/main/kotlin/io/rover/campaigns/core/events/UserInfo.kt [hunk 2]
```diff
@@ -1,44 +1,201 @@
...
             null -> hashMapOf()
-            else -> JSONObject(store[USER_INFO_KEY]).toAttributesHash()
+            else -> {
+                val info = JSONObject(store[USER_INFO_KEY]).toAttributesHash().toMutableMap()
+                if (info.containsKey("tags")) {
+                    log.v("Migrating user info tags")
+                    info["tags"]?.let { migrateTags(it) }
+                    info.remove("tags")
+                    store[USER_INFO_KEY] = info.encodeJson().toString()
+                    log.v("Migrated user info tags successfully")
```
Comment (samwoodall) [inline core/src/main/kotlin/io/rover/campaigns/core/events/UserInfo.kt:120]: Did you want to leave these logs here?

File: core/src/main/kotlin/io/rover/campaigns/core/events/UserInfo.kt [hunk 3]
```diff
@@ -1,44 +1,201 @@
...
+
+    override fun removeTag(tag: String) {
+        val mutableTags = currentTags.toMutableSet()
+        mutableTags.remove(Tag(value = tag, expires = null))
+        currentTags = mutableTags
+    }
+
+    override fun tags(): Set<String> {
+        return currentTags.map { it.value }.toSet()
+    }
```
[the remark this answers is not in the document]
Comment (chrisrecalis) [reply core/src/main/kotlin/io/rover/campaigns/core/events/UserInfo.kt:74]: whats standard?

File: core/src/main/kotlin/io/rover/campaigns/core/events/UserInfo.kt [hunk 4]
```diff
@@ -1,44 +1,201 @@
...
     override fun clear() {
         currentUserInfo = hashMapOf()
+        currentTags = setOf()
+    }
+
+    private var currentTags: Set<Tag> = try {
+        when (store[TAGS_KEY]) {
+            null -> setOf()
+            else -> decodeTags(store[TAGS_KEY]!!)
+        }
```
[the remark this answers is not in the document]
Comment (chrisrecalis) [reply core/src/main/kotlin/io/rover/campaigns/core/events/UserInfo.kt:85]: oh cool this is what I was looking for

File: core/src/main/kotlin/io/rover/campaigns/core/events/UserInfo.kt [hunk 5]
```diff
@@ -1,44 +1,201 @@
...
+        other as Tag
+
+        if (value != other.value) return false
+
+        return true
+    }
+
+    override fun hashCode(): Int {
+        return value.hashCode()
+    }
```
[the remark this answers is not in the document]
Comment (chrisrecalis) [reply core/src/main/kotlin/io/rover/campaigns/core/events/UserInfo.kt:180]: Yeah two tags with the same value but different expires. are considered the same. However whenever adding to the set of tags it should overwrite the existing value. Is there a better data structure for this? I haven't checked to see if calling `add` actually overwrites the existing value I'm going to assume not...

File: core/src/main/kotlin/io/rover/campaigns/core/events/UserInfo.kt (hunk 2 above)
Comment (chrisrecalis) [reply core/src/main/kotlin/io/rover/campaigns/core/events/UserInfo.kt:120]: idk should i? lol 😂. I can remove it.

File: core/src/main/kotlin/io/rover/campaigns/core/events/UserInfo.kt [hunk 6]
```diff
@@ -1,44 +1,201 @@
...
+                    else -> it.expires.after(now)
+                }
+            }.toSet()
+
+            if (unExpiredTags != field) {
+                // Some tags have expired update the value
+                currentTags = unExpiredTags
+            }
+            return unExpiredTags
+        }
```
[the remark this answers is not in the document]
Comment (chrisrecalis) [reply core/src/main/kotlin/io/rover/campaigns/core/events/UserInfo.kt:104]: Question about naming. `isExpired` vs `hasExpired` function?

File: core/src/main/kotlin/io/rover/campaigns/core/events/UserInfo.kt (hunk 2 above)
Comment (samwoodall) [reply core/src/main/kotlin/io/rover/campaigns/core/events/UserInfo.kt:120]: Yeah, i'd remove it, I don't think it adds a ton of value

File: core/src/main/kotlin/io/rover/campaigns/core/events/UserInfo.kt (hunk 6 above)
[the remark this answers is not in the document]
Comment (samwoodall) [reply core/src/main/kotlin/io/rover/campaigns/core/events/UserInfo.kt:104]: I prefer `isExpired` it's more consistent with other similar funcs too

File: core/src/main/kotlin/io/rover/campaigns/core/events/UserInfo.kt (hunk 3 above)
[the remark this answers is not in the document]
Comment (samwoodall) [reply core/src/main/kotlin/io/rover/campaigns/core/events/UserInfo.kt:74]: It depends on whether you can easily tell the return type, I don't think I can in this case, so keep it explicitly stated

**Comment by chrisrecalis** [comment]
@samwoodall take a look at my model changes. I decided to go with a private data class as I think it's easier to encapsulate the expiring functionality.

**Pull request closed without merging**